tests/client/conftest.py

```diff
@@ -0,0 +1,145 @@
+from contextlib import asynccontextmanager
+from unittest.mock import patch
+
+import pytest
+
+import mcp.shared.memory
+from mcp.shared.message import SessionMessage
+from mcp.types import (
+    JSONRPCNotification,
+    JSONRPCRequest,
+)
+
+
+class SpyMemoryObjectSendStream:
+    def __init__(self, original_stream):
+        self.original_stream = original_stream
+        self.sent_messages: list[SessionMessage] = []
+
+    async def send(self, message):
+        self.sent_messages.append(message)
+        await self.original_stream.send(message)
+
+    async def aclose(self):
+        await self.original_stream.aclose()
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, *args):
+        await self.aclose()
+
+
+class StreamSpyCollection:
+    def __init__(
+        self,
+        client_spy: SpyMemoryObjectSendStream,
+        server_spy: SpyMemoryObjectSendStream,
+    ):
+        self.client = client_spy
+        self.server = server_spy
+
+    def clear(self) -> None:
+        """Clear all captured messages."""
+        self.client.sent_messages.clear()
+        self.server.sent_messages.clear()
+
+    def get_client_requests(self, method: str | None = None) -> list[JSONRPCRequest]:
+        """Get client-sent requests, optionally filtered by method."""
+        return [
+            req.message.root
+            for req in self.client.sent_messages
+            if isinstance(req.message.root, JSONRPCRequest)
+            and (method is None or req.message.root.method == method)
+        ]
+
+    def get_server_requests(self, method: str | None = None) -> list[JSONRPCRequest]:
+        """Get server-sent requests, optionally filtered by method."""
+        return [
+            req.message.root
+            for req in self.server.sent_messages
+            if isinstance(req.message.root, JSONRPCRequest)
+            and (method is None or req.message.root.method == method)
+        ]
+
+    def get_client_notifications(
+        self, method: str | None = None
+    ) -> list[JSONRPCNotification]:
+        """Get client-sent notifications, optionally filtered by method."""
+        return [
+            notif.message.root
+            for notif in self.client.sent_messages
+            if isinstance(notif.message.root, JSONRPCNotification)
+            and (method is None or notif.message.root.method == method)
+        ]
+
+    def get_server_notifications(
+        self, method: str | None = None
+    ) -> list[JSONRPCNotification]:
+        """Get server-sent notifications, optionally filtered by method."""
+        return [
+            notif.message.root
+            for notif in self.server.sent_messages
+            if isinstance(notif.message.root, JSONRPCNotification)
+            and (method is None or notif.message.root.method == method)
+        ]
+
+
+@pytest.fixture
+def stream_spy():
+    """Fixture that provides spies for both client and server write streams.
```

Comment:
Built this as a fixture so that it can be easily reused in multiple tests.

I needed a way to verify the request payload "on the wire" and this seemed like a good approach, since it did not require any modification of the actual client code (only test fixtures).